plugin/secrets_service_account_key.go

```diff
@@ -213,13 +297,90 @@ func (b *backend) getSecretKey(ctx context.Context, s logical.Storage, rs *RoleS
 	}
 
 	resp := b.Secret(SecretTypeKey).Response(secretD, internalD)
-	resp.Secret.Renewable = true
+	resp.Secret.Renewable = false
+	resp.Secret.MaxTTL = cfg.MaxTTL
+	resp.Secret.TTL = ttlToUse
+
+	return resp, nil
+}
+
+func (b *backend) constructRespFromCache(item *serviceAccountKeyCacheItem, ttl, maxTTL time.Duration) *logical.Response {
+	secretD, internalD := item.secretResponse()
+	resp := b.Secret(SecretTypeKey).Response(secretD, internalD)
+	resp.Secret.TTL = ttl
+	resp.Secret.MaxTTL = maxTTL
+	resp.Secret.Renewable = false
 
-	if ttl > 0 {
-		resp.Secret.TTL = time.Duration(ttl) * time.Second
+	return resp
+}
+
+func getSecretKeyFromCache(ctx context.Context, s logical.Storage, rs *RoleSet) (*serviceAccountKeyCacheItem, error) {
+	cacheCollection, err := getCacheCollection(ctx, s, rs.Name)
+	if err != nil {
+		return nil, errwrap.Wrapf("could not get service account key cache collection: {{err}}", err)
 	}
 
-	return resp, nil
+	if cacheCollection != nil {
+		_, validCacheItem := cacheCollection.getFirstValidItem(rs.bindingHash())
+		if validCacheItem != nil {
+			validCacheItem.Counter++
+
+			if err := cacheCollection.putToStorage(ctx, s, rs.Name); err != nil {
+				return nil, errwrap.Wrapf("failed to update cache key collection: {{err}}", err)
+			}
+
+			return validCacheItem, nil
+		}
+	}
+
+	return nil, nil
```

Comment:
I think this block can be improved this way.

```
	if cacheCollection == nil {
		return nil, nil
	}

	_, validCacheItem := cacheCollection.getFirstValidItem(rs.bindingHash())
	if validCacheItem == nil {
		return nil, nil
	}

	validCacheItem.Counter++
	if err := cacheCollection.putToStorage(ctx, s, rs.Name); err != nil {
		return nil, errwrap.Wrapf("failed to update cache key collection: {{err}}", err)
	}

	return validCacheItem, nil
```

Reply:
I think that's better. I'll update this block in the next commit. Thank you.